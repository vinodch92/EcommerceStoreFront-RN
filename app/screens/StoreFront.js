import React from 'react';
import { StyleSheet, SafeAreaView, FlatList } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { ProductCard } from '../components';
import { ProductsState } from '../state';

const StoreFront = () => {
  const dispatch = useDispatch();

  const products = useSelector(ProductsState.selectProductsList);

  React.useEffect(() => {
    callProductsListApi();
  }, []);

  const callProductsListApi = () => dispatch({ type: 'GET_PRODUCTS_LIST', payload: '' });

  const renderCard = ({ item }) => <ProductCard productData={item} />;

  return (
    <SafeAreaView style={styles.layout}>
      <FlatList
        keyboardShouldPersistTaps="handled"
        data={products}
        keyExtractor={(item, index) => index.toString()}
        renderItem={renderCard}
      />
    </SafeAreaView>
  );
};

export default StoreFront;

const styles = StyleSheet.create({
  layout: { flex: 1 },
});
