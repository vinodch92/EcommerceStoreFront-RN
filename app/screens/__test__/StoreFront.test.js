import React from 'react';
import { shallow } from 'enzyme';
import { FlatList, SafeAreaView } from 'react-native';
import * as redux from 'react-redux';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';
import StoreFront from '../StoreFront';

const mockProducts = [
  {
    name: 'Brown eggs',
    image:
      'https://raw.githubusercontent.com/vinodch92/ECommerceProductImages/master/images/storeFront/0.jpg',
    rating: 4,
  },
  {
    name: 'Sweet fresh stawberry',
    image:
      'https://raw.githubusercontent.com/vinodch92/ECommerceProductImages/master/images/storeFront/1.jpg',
    rating: 4,
  },
  {
    name: 'Asparagus',
    image:
      'https://raw.githubusercontent.com/vinodch92/ECommerceProductImages/master/images/storeFront/2.jpg',
    rating: 3,
  },
];

describe('StoreFront', () => {
  const dispatch = jest.fn();
  redux.useDispatch.mockReturnValue(dispatch);

  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware]);

  let useEffect;
  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((fn) => fn());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect();
  });

  describe('Rendering', () => {
    const store = mockStore({
      products: mockProducts,
    });

    jest.spyOn(redux, 'useSelector').mockImplementationOnce(() => store.getState().products);
    const wrapper = shallow(<StoreFront />);
    it('renders component correctly', () => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find(SafeAreaView).length).toEqual(1);
      expect(wrapper.find(FlatList).length).toEqual(1);

      const flatList = wrapper.find(FlatList);
      flatList.renderProp('renderItem')({ item: mockProducts[0] }, 0);
    });
  });
});
