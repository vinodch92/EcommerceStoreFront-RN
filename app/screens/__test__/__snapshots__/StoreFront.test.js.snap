// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreFront Rendering renders component correctly 1`] = `ShallowWrapper {}`;
